src/Crypto/WebAuthn/Operation/Authentication.hs

```diff
@@ -215,55 +215,49 @@ verifyAuthenticationResponse origin rpIdHash midentifiedUser entry options crede
   let allowCredentials = M.coaAllowCredentials options
   unless
     (null allowCredentials || M.cIdentifier credential `elem` map M.cdId allowCredentials)
-    . failure
+    . throwError
     $ AuthenticationCredentialDisallowed allowCredentials credential
 
   -- 6. Identify the user being authenticated and verify that this user is the
   -- owner of the public key credential source credentialSource identified by
   -- credential.id:
-  --
-  -- -> If the user was identified before the authentication ceremony was
-  -- initiated, e.g., via a username or cookie, verify that the identified
-  -- user is the owner of credentialSource. If response.userHandle is present,
-  -- let userHandle be its value. Verify that userHandle also maps to the same
-  -- user.
-  --
-  -- -> If the user was not identified before the authentication ceremony was
-  -- initiated, verify that response.userHandle is present, and that the user
-  -- identified by this value is the owner of credentialSource.
-  let owner = ceUserHandle entry
-
-  -- According to the [(spec)](https://www.w3.org/TR/webauthn-2/#dom-publickeycredentialuserentity-id)
-  -- The user handle MUST NOT be empty, though it MAY be null.
-  -- For clarification see https://github.com/w3c/webauthn/issues/1722
-  -- However, Safari returns an empty string instead of null, see the bug report:
-  -- https://bugs.webkit.org/show_bug.cgi?id=239737
-  let mUserHandler = case M.araUserHandle response of
-        Just (M.UserHandle "") -> Nothing
-        userHandle -> userHandle
-
-  case (midentifiedUser, mUserHandler) of
-    (Just identifiedUser, Just userHandle)
-      | identifiedUser /= owner ->
-          failure $ AuthenticationIdentifiedUserHandleMismatch identifiedUser owner
-      | userHandle /= owner ->
-          failure $ AuthenticationCredentialUserHandleMismatch userHandle owner
-      | otherwise -> pure ()
-    (Just identifiedUser, Nothing)
-      | identifiedUser /= owner ->
-          failure $ AuthenticationIdentifiedUserHandleMismatch identifiedUser owner
-      | otherwise -> pure ()
-    (Nothing, Just userHandle)
-      | userHandle /= owner ->
-          failure $ AuthenticationCredentialUserHandleMismatch userHandle owner
-      | otherwise -> pure ()
-    (Nothing, Nothing) ->
-      failure AuthenticationCannotVerifyUserHandle
+  (credentialRecord, userHandle) <- case userIdentified of
```

Comment:
I like this change; as now the algorithm actually looks like how the spec is written


> Identify the user being authenticated and verify that this user is the owner of the [public key credential source](https://www.w3.org/TR/webauthn-2/#public-key-credential-source) credentialSource identified by credential.[id](https://w3c.github.io/webappsec-credential-management/#dom-credential-id):
> 
> If the user was identified before the [authentication ceremony](https://www.w3.org/TR/webauthn-2/#authentication-ceremony) was initiated, e.g., via a username or cookie,
verify that the identified user is the owner of credentialSource. If response.[userHandle](https://www.w3.org/TR/webauthn-2/#dom-authenticatorassertionresponse-userhandle) is present, let userHandle be its value. Verify that userHandle also maps to the same user.
> 
> If the user was not identified before the [authentication ceremony](https://www.w3.org/TR/webauthn-2/#authentication-ceremony) was initiated,
verify that response.[userHandle](https://www.w3.org/TR/webauthn-2/#dom-authenticatorassertionresponse-userhandle) is present, and that the user identified by this value is the owner of credentialSource.
https://www.w3.org/TR/webauthn-2/#sctn-verifying-assertion